### PR TITLE
Legend bounding box update fix

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

Fixing the legend offset problem introduced in https://github.com/recharts/recharts/pull/4273

The problem is that after my changes, the component that reads and updates the `boundingClientRect` and the component that re-renders when payload changes, are no longer the same component.

In this PR I have moved both the bounding box, plus payload read, to the same component so that it can call `onBBoxUpdate` again.

In my opinion the `onBBoxUpdate` should not exist in the first place because all it does is it passes the width and height back to Legend as `width` and `height` props. But that's a refactor for another day.

## Related Issue

https://github.com/recharts/recharts/pull/4273

## Motivation and Context

Fixing 3.x regression

## How Has This Been Tested?

npm test
storybook

## Screenshots (if appropriate):
image
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [x] All new and existing tests passed.

### CODE REVIEW

File: test/component/Legend.spec.tsx [hunk 1]
```diff
@@ -1339,7 +1426,6 @@ describe('<Legend />', () => {
         align: LegendProps['align'];
         verticalAlign: LegendProps['verticalAlign'];
         layout: LegendProps['layout'];
-        expectedStyle: string;
```
Comment (PavelVanecek) [inline test/component/Legend.spec.tsx:1342]: As a bonus - the Legend now renders with the final styles, right away. While before it required render, then read from DOM, and adjustment on second render.

File: test/helper/mockGetBoundingClientRect.ts [hunk 2]
```diff
@@ -17,10 +18,17 @@ export function mockGetBoundingClientRect(rect: Partial<DOMRect>) {
     right: 0,
     bottom: 0,
     left: 0,
-    toJSON: () => '{}',
     ...rect,
+  };
+  Element.prototype.getBoundingClientRect = () => ({
+    ...mockDomRect,
+    toJSON: () => JSON.stringify(mockDomRect),
```
Comment (PavelVanecek) [inline test/helper/mockGetBoundingClientRect.ts:25]: Vitest will use this method to print difference if a test fails. So having the hardcoded `{}` was not helping.

**Review by ckifer — APPROVED** [review]
Looks great! Thanks for the fix

File: test/helper/mockGetBoundingClientRect.ts (hunk 2 above)
Comment (ckifer) [reply test/helper/mockGetBoundingClientRect.ts:25]: who knew, I'll put that one in my back pocket for later

**Pull request merged**